File: javascript/206.reverse-linked-list.js
/*
 * @lc app=leetcode id=206 lang=javascript
 *
 * [206] Reverse Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
  // Solution 1:
  // let revHead = null;
  // while (head) {
  //   const temp = head;
  //   head = head.next;
  //   temp.next = revHead;
  //   revHead = temp;
  // }
  // return revHead;

  // Solution 2: recursion
  let revHead = head;
  if (head && head.next) {
    revHead = reverseList(head.next);
    head.next.next = head;
    head.next = null;
  }
  return revHead;
};
// @lc code=end

File: rust/206.reverse-linked-list.rs
/*
 * @lc app=leetcode id=206 lang=rust
 *
 * [206] Reverse Linked List
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let (mut h, mut r) = (head, None);
        while let Some(mut node) = h {
            h = node.next;
            node.next = r;
            r = Some(node);
        }
        return r;
    }
}
// @lc code=end
